File: Algo/Hard/0239_Sliding_Window_Maximum/Solution.cpp
// 2022-10-13 (ordered map)
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> output;
        map<int, int> contains;
        int l = 0, r = 0, n = nums.size();

        while (r < n) {
            if (!contains.count(nums[r]))
                contains[nums[r]] = 0;
            contains[nums[r]]++;
            r++;
            
            if (r-l == k) {
                auto back = contains.rbegin();
                output.push_back(back->first);
                contains[nums[l]] -= 1;
                if (contains[nums[l]] == 0)
                    contains.erase(nums[l]);
                l++;
            }
        }

        return output;
    }
};

File: Algo/Hard/0239_Sliding_Window_Maximum/Solution.py
from collections import deque
from typing import *

## 2022-10-22 (monotonic deque)
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        window = deque()
        output = []
        
        for i, val in enumerate(nums):
            while window and window[0]+k <= i:
                window.popleft()
            
            while window and nums[window[-1]] < val:
                window.pop()
            
            window.append(i)

            if i >= k-1:
                output.append(nums[window[0]])
        
        return output
            